test/test_stone.py

```diff
@@ -1315,6 +1445,340 @@ def test_forward_reference(self):
+                example default
+                    user_id="1234"
+                example pro
+                    user_id="1234P"
+            patch struct QuotaInfo
+                quota UInt64
+                    "The user's total quota allocation (bytes)."
+                example default
+                    quota=2000000000
+                example pro
+                    quota=100000000000
+            """)
+        api = specs_to_ir([('test.stone', text)])
+
+        quota_info_dt = api.namespaces['files'].data_type_by_name['QuotaInfo']
+        self.assertEqual(quota_info_dt.fields[1].name, 'quota')
+        self.assertTrue(is_integer_type(quota_info_dt.fields[1].data_type))
+        self.assertEqual(quota_info_dt.fields[1].doc, "The user's total quota allocation (bytes).")
+        self.assertEqual(quota_info_dt.get_examples()['default'].value['quota'], 2000000000)
+        self.assertEqual(quota_info_dt.get_examples()['pro'].value['quota'], 100000000000)
+
+        # Test patching inherited struct
+        text = textwrap.dedent("""\
+            namespace files
+            struct QuotaInfo
+                "The space quota info for a user."
+                user_id String
+                    "The user associated with the quota."
+                example default
+                    user_id="1234"
+                example pro
+                    user_id="1234P"
+            struct QuotaInfoPersonal extends QuotaInfo
+                "The space quota info for a personal user."
+                personal_quota UInt64
+                    "The user's personal quota allocation (bytes)."
+            patch struct QuotaInfo
+                quota UInt64
+                    "The user's total quota allocation (bytes)."
+                example default
+                    quota=2000000000
+                example pro
+                    quota=100000000000
+            """)
+        api = specs_to_ir([('test.stone', text)])
+
+        quota_info_dt = api.namespaces['files'].data_type_by_name['QuotaInfoPersonal']
+        self.assertEqual(quota_info_dt.all_fields[1].name, 'quota')
+        self.assertTrue(is_integer_type(quota_info_dt.all_fields[1].data_type))
+        self.assertEqual(
+            quota_info_dt.all_fields[1].doc, "The user's total quota allocation (bytes).")
+
+        # Test patching enumerated supertype struct
+        text = textwrap.dedent("""\
+            namespace test
+            struct Resource
+                union
+                    file File
+                    folder Folder
+            struct File extends Resource
+                size UInt64
+                    "The size of the file."
+                example default
+                    size=5
+            struct Folder extends Resource
+                icon String
+                    "The name of the icon."
+                example default
+                    icon="My Icon"
+            patch struct Resource
```

Comment:
What if I tried to declare a new union subtype here?

Reply:
This fails in the parsing layer, albeit with a rather unclear message. I added a test. Look for `# test no patching enumerated subtype`.